Add minimap to TextEdit

Paulb23 opened the pull request:
Adds an optional text minimap to `text_edit` similar to that of sublime text.

It supports wrapping, code folding, and scrolling. In addition it uses full syntax highlighting, current line and viewport outline. However does not highlight beyond that eg searches.  It currently draws the characters as blocks.

Due to the extra processing, there was a significant performance hit around syntax highlighting, to combat this `text_edit` will now cache syntax highlighting colours similar to that of colour regions. 

Preview:

![image](https://user-images.githubusercontent.com/6584330/62838137-14a18700-bc70-11e9-87d7-7f9e31e8a63d.png)

closes #4881

Paulb23 committed:
Add syntax highlighting cache

Paulb23 committed:
Add minimap to text_edit

Calinou commented:
> It currently draws the characters as blocks.

You can make blocks look closer to actual characters by toning down their opacity slightly (try 70% opacity).

YeldhamDev commented:
I think the width should be exposed to the settings, and it's default should be the equivalent of the default line guideline (80).

Zylann commented:
I have attempted something similar before, so here are my findings:

- You can make the minimap look better by drawing textured rectangles instead of full blocks (the texture being just 1x2 with the top pixel half-transparent), and draw them without gap to fit more lines, like so:
![image](https://user-images.githubusercontent.com/1311555/62842710-da55db00-bcab-11e9-8491-b64b58c07a32.png)

- I did some benchmarking for my C++ version and quickly found that even when caching syntax coloring into an O(1) container (right next to characters, really), the drawing operation was the slowest part, just enough to almost halve framerate compared to not drawing the minimap (to the point I ditched Godot for that project and use SDL instead, which was twice as fast). The Godot editor is already quite crowded, a minimap can add up a lot. Did you check draw time and framerate?

- You are drawing every character individually, which may be even slower than what I attempted because it multiplies draw calls. I found batching similarly-coloured blocks together brings improvements.

Calinou commented:
@Zylann I guess this feature should be disabled by default until the performance bottlenecks are solved. To me, it seems better to have a smooth editor without a minimap compared to a sluggish editor with a minimap.

Paulb23 commented:
@Zylann Just did a single run benchmark using `OS::get_singleton()->get_ticks_usec();` at the start and end of `text_edits` draw notification.

Here's the mean results:

- No Highlighter Cache, Minimap Diabaled (As it is currently): 8326.3560277537
- No Highlighter Cache, Minimap Enabled: 14210.949481865
- Highlighter Cache, Minimap Disabled: 6660.104175635
- Highlighter Cache, Minimap Enabled: 9555.1651455889

Paulb23 commented:
Added a setting for the minimap width and it now slightly fades out the font colours. preview:

![image](https://user-images.githubusercontent.com/6584330/62904241-68d66500-bd5d-11e9-9e14-edc8a9f058d2.png)

YeldhamDev commented:
By the way, I think the options should be exposed to scripting.

Paulb23 commented:
* Fixed syntax highlighter cache not clearing properly.
* Fixed minimap syntax highlighting when wrapping.
* Fixed minimap initial indent when wrapping.
* The minimap will now indent more for tabs then spaces.
* Exposed to scripting

The minimap now draws blocks of the same colours rather then one char at a time, resulting is a small performance boost  from from 9555 to around 9000 mean.

akien-mga commented:
Thanks!

akien-mga commented:
Shouldn't we enable the minimap by default? Otherwise most users won't even notice that the feature exists.
